### PR TITLE
Options hookup

### PR DESCRIPTION
**Pull request by pq**

### CODE REVIEW

**Comment by pq** [comment]
More hardening (and testing) to come but this will get us to where we can play with client integration (e.g., the linter).

@bwilkerson PTAL

File: lib/src/driver.dart [hunk 1]
```diff
@@ -59,15 +61,27 @@ class Driver {
...
+    // Parse commandline options.
     CommandLineOptions options = CommandLineOptions.parse(args);
-    // In batch mode, SDK is specified on the main command line rather than in
-    // the command lines sent to stdin.  So process it before deciding whether
-    // to activate batch mode.
-    if (sdk == null) {
-      sdk = new DirectoryBasedDartSdk(new JavaFile(options.dartSdkPath));
+
+    // Set up env.
+    {
```
Comment (bwilkerson) [inline lib/src/driver.dart]: I don't understand the value of the added block, but I guess I don't care if you leave it.

File: lib/src/driver.dart [hunk 2]
```diff
@@ -324,11 +338,35 @@ class Driver {
...
+        AnalysisEngine.instance.optionsPlugin.optionsProcessors;
+
+    // Read file and notify processors
+    try {
+      String contents = file.readAsStringSync();
+      Map<String, YamlNode> options = new OptionsFileParser().parse(contents);
+      optionsProcessors
+          .forEach((OptionsProcessor p) => p.optionsProcessed(options));
+    } catch (e) {
+      optionsProcessors.forEach((OptionsProcessor p) => p.onError(e));
```
Comment (bwilkerson) [inline lib/src/driver.dart:354]: The method 'onError' is typed to take an 'Exception', but 'e' can be anything. We should add a type to the catch clause.

File: lib/src/options.dart [hunk 3]
```diff
@@ -436,3 +447,51 @@ class CommandLineParser {
...
+///     bool useMultiPackage =
+///         options['compiler']['resolver']['useMultiPackage'];
+class OptionsFileParser {
+  /// Parse the given [yaml] source into a mapping of option scopes to
+  /// associated option content [YamlNode]s.
+  Map<String, YamlNode> parse(String yaml) {
+    var options = <String, YamlNode>{};
+    var doc = loadYaml(yaml);
+    if (doc is! YamlMap) {
+      throw 'Bad options file format (expected map, got ${doc.runtimeType})';
```
Comment (bwilkerson) [inline lib/src/options.dart]: Personally, I consider it to be bad form to throw something other than an exception, and it violates the contract we set up with OptionsProcessor. (Both here and below.)

File: pubspec.yaml [hunk 4]
```diff
@@ -3,15 +3,16 @@ version: 1.1.0
 description: Command line interface for the Dart Analyzer.
 author: Dart Team <misc@dartlang.org>
 homepage: https://github.com/dart-lang/analyzer_cli
+environment:
+  sdk: '>=1.8.0 <2.0.0'
```
Comment (bwilkerson) [inline pubspec.yaml:7]: Should this be "^1.8.0"?

**Comment by bwilkerson** [comment]
LGTM

File: lib/src/driver.dart (hunk 2 above)
Comment (pq) [inline lib/src/driver.dart:354]: :+1:

File: lib/src/driver.dart (hunk 1 above)
Comment (pq) [inline lib/src/driver.dart]: Full disclosure: I think this is the first time I've done this.  Ever.  I believe Konstantin uses this idiom?  I'm playing with ways to make things more readable.  My initial temptation was to create a new method (a la `_processPlugins()`) but that seemed a bit much.  Now that I look at it though, a method feels fine after all.  Updating...

File: lib/src/options.dart (hunk 3 above)
Comment (pq) [inline lib/src/options.dart]: Will fix.

File: pubspec.yaml (hunk 4 above)
Comment (pq) [inline pubspec.yaml:7]: You'd think.  It's a bootstrapping thing.  They could have done this differently but the idea is that the one place you can't use these clever constraints is if you're testing if you _can_ use these smart constraints (e.g., `sdk >= 1.8.0`).

**Pull request merged**